### PR TITLE
Add support with multi-gpu train for train_newtork.py

### PR DESCRIPTION
**Pull request by Isotr0py**
Support DistributedDataParallel training for `train_newtork.py`.
Now `accelerate launch` with `--multi_gpu` argument can enable DDP training with multiple GPUs.

I've tested `train_newtork.py` with two Tesla T4 GPUs and tried to fix some bugs appeared in DDP training.
`unet`, `text_encoder`, `network` should be extracted from class `DistributedDataParallel` through attribute `.module` in DDP training.

I haven't used `train_db.py` and `train_TI.py` and didn't make any changes to them. I will test them later.

Some unnecessary code was inserted just to reduce RAM usage for my poor low-ram device, please ignore it, XD.

XLAP)% MD7YV )%SJKH@Z0C

### CODE REVIEW

**Review by sheldonchiu — CHANGES_REQUESTED** [review]
```python
from torch.cuda.amp import autocast
from torch.nn.parallel import DistributedDataParallel as DDP
```
```python
  text_encoder.requires_grad_(False)
  text_encoder.to(accelerator.device)
  if args.gradient_checkpointing:                       # according to TI example in Diffusers, train is required
    unet.train()
    text_encoder.train()

    # set top parameter requires_grad = True for gradient checkpointing works
    if type(text_encoder) == DDP:
      text_encoder.module.text_model.embeddings.requires_grad_(True)
    else:
      text_encoder.text_model.embeddings.requires_grad_(True)
  else:
    unet.eval()
    text_encoder.eval()

  # support DistributedDataParallel
  if type(text_encoder) == DDP:
      text_encoder = text_encoder.module
      unet = unet.module
      network = network.module
```
Doing a type checking should be a better method, and this little fix is need for gradient checkpointing to work.

**Comment by estaesta** [comment]
how do you manage to run using multi gpu on Kaggle? I've updated my sd-script but it is still crashing because of the RAM

**Comment by Isotr0py** [comment]
> how do you manage to run using multi gpu on Kaggle? I've updated my sd-script but it is still crashing because of the RAM

I have set my training notebook public in Kaggle. Maybe it can help you.
https://www.kaggle.com/code/isotr0py/lora-train-multi-gpus

**Comment by Penguin0321** [comment]
I wonder how do you pass the parameter "multi_gpu" in a certain function since I don't see any "multi_gpu" in code

**Comment by Isotr0py** [comment]
`--multi_gpu` is the hardware selection argument from `accelerate launch` instead of the script.
You can find reference from their document: [using-accelerate-launch](https://huggingface.co/docs/accelerate/v0.14.0/en/basic_tutorials/launch#using-accelerate-launch)
